pkg/reconciler/v1alpha1/autoscaling/kpa_scaler.go

```diff
@@ -84,6 +86,46 @@ func (ks *kpaScaler) getAutoscalerConfig() *autoscaler.Config {
 	return ks.autoscalerConfig.DeepCopy()
 }
 
+func scaleBoundInt32(ctx context.Context, kpa *kpa.PodAutoscaler, key string) int32 {
+	logger := logging.FromContext(ctx)
+	if s, ok := kpa.Annotations[key]; ok {
+		i, err := strconv.ParseInt(s, 10, 32)
+		if err != nil || i < 0 {
+			logger.Debugf("value of %s must be positive integer", key)
+			return 0
+		}
+		return int32(i)
+	}
+	return 0
+}
+
+func getScaleBounds(ctx context.Context, kpa *kpa.PodAutoscaler) (int32, int32) {
+	logger := logging.FromContext(ctx)
+
+	min := scaleBoundInt32(ctx, kpa, autoscaling.MinScaleAnnotationKey)
+	max := scaleBoundInt32(ctx, kpa, autoscaling.MaxScaleAnnotationKey)
+
+	if max != 0 && min > max {
+		logger.Debugf("%s must be less than or equal %s, adjusting min = max", autoscaling.MinScaleAnnotationKey, autoscaling.MaxScaleAnnotationKey)
+		min = max
+	}
+
+	return min, max
+}
+
+// pre: 0 <= min <= max && 0 <= x
+func applyBounds(min, max int32) func(int32) int32 {
+	return func(x int32) int32 {
+		if min > 0 && x < min {
+			return min
+		}
+		if max > 0 && x > max {
```

Comment:
Since `scaleBoundInt32` always returns a non-negative value, this test is equivalent to:
```
if max != 0 && x > max {
```

This leaves me wondering if `max == 0` is being used to indicate "no upper bound". If so, wouldn't it be better to indicate "no upper bound" by setting `max` to the maximal `int32` (internally - I'm comfortable with the user using '0' to indicate "no upper bound" _providing_ that is documented) and then the test could become the more obvious:
```
if x > max {
```
?

Reply:
@imikushin Yeah, that was my point. Why not represent the user not having set an upper bound explicitly by setting `max` to the maximal `int32` so that this test becomes simpler? I'm trying to separate the user interface (special use of `0`) from the semantics of maximum value. It's not a deal breaker though if you don't agree. :-)

Reply:
Thanks for your comments @glyn!